lib/migrate.js

```diff
@@ -0,0 +1,448 @@
+      var results = {};
+      for (var key in indexes) {
+        var keys = indexes[key].keys;
+        assert(keys && typeof keys === 'object',
+          'the keys in your model index is not well defined!' +
+          'please check documentation: define index in loopback model');
+        var fields = [];
+        _.forEach(keys, function(value, key) {
+          var obj = {};
+          var order;
+          if (keys[key] === 1) order = 'asc';
+          else order = 'desc';
+          obj[key] = order;
+          fields.push(obj);
+        });
+        results[key] = fields;
+      }
+      return results;
+    }
+  };
+
+  /**
+ * Perform the indexes comparison for `autoupdate`.
+ * @param {Object} newIndexes
+ * newIndexes in format:
+ * ```
+ * {
+ *   indexName: [{afield: 'asc'}],
+ *   compositeIndexName: [{field1: 'asc'}, {field2: 'asc'}]
+ * }
+ * ```
+ * @param {Object} oldIndexes
+ * oldIndexes in format:
+ * ```
+ * {
+ *   indexName: {
+ *     ddoc: '_design/LBModel__Foo__LBIndex__bar_index',
+ *     fields: [{afield: 'asc'}],
+ *   }
+ * }
+ * ```
+ * @callback {Function} cb The callback function
+ * @param {Object} result indexes to add and drop after comparison
+ * - result: {indexesToAdd: {$someIndexes}, indexesToDrop: {$someIndexes}}
+ */
+  CouchDB.prototype.compare = function(newIndexes, oldIndexes, cb) {
+    debug('CouchDB.prototype.compare');
+    var result = {};
+    var indexesToDrop = {};
+    var indexesToAdd = {};
+    var iAdd;
+    for (var niKey in newIndexes) {
+      if (!oldIndexes.hasOwnProperty(niKey)) {
+      // Add item to `indexesToAdd` if it's new
+        iAdd = {};
+        iAdd[niKey] = newIndexes[niKey];
+        indexesToAdd = _.merge(indexesToAdd, iAdd);
+      } else {
+        if (arrEqual(newIndexes[niKey], oldIndexes[niKey].fields)) {
+        // Don't change it if index already exists
+          delete oldIndexes[niKey];
+        } else {
+        // Update index if fields change
+          iAdd = {};
+          var iDrop = {};
+          iAdd[niKey] = newIndexes[niKey];
+          indexesToAdd = _.merge(indexesToAdd, iAdd);
+        }
+
+        // function arrEqual(arr1, arr2) {
+        //   return _.isEqual(arr1.sort(), arr2.sort());
+        // }
+        // IMPROVE: use _.isEqual with customized compare function?
+        // index fields array is in format: [{afield: 'asc'}, {anotherfield: 'asc'}]
+        function arrEqual(arr1, arr2) {
+          assert(Array.isArray(arr1) && Array.isArray(arr2), 'make sure you pass in two Array!');
+          var notEqualMsg = util.inspect(arr1, 4) + ' is not equal to ' +
+        util.inspect(arr2, 4);
+          var isEqual = true;
+          for (var item in arr1) {
+            var i = _.findIndex(arr2, arr1[item]);
+            isEqual = isEqual && (i > -1);
+          }
+          return isEqual;
+        }
+      }
+    }
+    indexesToDrop = oldIndexes;
+    // Important, make sure `indexesToAdd` and `indexesToDrop` are object
```

Comment:
I don't think we should be using comments to do what tests are meant to do. If someone removes these lines of code and tests don't immediately begin bursting into flames, then comments won't help us much either.